### PR TITLE
planner: refine Cardinality to map

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

The [`Cardinality`](https://github.com/pingcap/tidb/blob/fe03864/planner/property/stats_info.go#L25) currently is defined in `[]float64` which the index of the slice is the index of the [`Columns`](https://github.com/pingcap/tidb/blob/fe03864/expression/schema.go#L37) in a [`Schema`](https://github.com/pingcap/tidb/blob/fe03864/expression/schema.go#L36).
But it's hard to maintain the `Cardinality` because whenever the order of `Columns` in a `Schema` is changed, or some `Columns` in a `Schema` are pruned, the `Cardinality` must be followed by the new changes of the index.

### What is changed and how it works?

1. refine the `Cardinality`, change from `[]float64` which the index of the slice is the index of the columns, to `map[int64]float64` as a map from [`Column.UniqueID`](https://github.com/pingcap/tidb/blob/fe03864/expression/column.go#L165) to `Cardinality`. After that, `Cardinality` will not be affected by the sequence changes of the `Columns`.
2. Luckily resolve an existing bug that the left and right columns of a schema are changed by `join reorder` but the cardinality is not followed. (Fix some TPCH explain results)

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

Release note

 - None

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -37,10 +37,10 @@ func (p *basePhysicalPlan) StatsCount() float64 {
 func (p *LogicalTableDual) DeriveStats(childStats []*property.StatsInfo, selfSchema *expression.Schema, childSchema []*expression.Schema) (*property.StatsInfo, error) {
 	profile := &property.StatsInfo{
 		RowCount:    float64(p.RowCount),
-		Cardinality: make([]float64, selfSchema.Len()),
+		Cardinality: make(map[int64]float64, selfSchema.Len()),
```
Comment (lamxTyler) [inline planner/core/stats.go]: Would it cause performance issues after using map?

Comment (SunRunAway) [reply planner/core/stats.go]: Probably, but IMO the cost of `map` from `slice` is not that significant and is worth handling code friendly in the future. (Let's do a `/bench`)

**Comment by eurekaka** [comment]
Shall we cherry-pick this PR to release branch since it contains a bug fix?

**Comment by SunRunAway** [comment]
> Shall we cherry-pick this PR to release branch since it contains a bug fix?

The 3.0 branch does not use cardinality heavily so I think we can keep it in 4.0

**Pull request merged**